crates/proto/src/rustls/tls_server.rs

```diff
@@ -35,29 +34,18 @@ pub fn read_cert(cert_path: &Path) -> ProtoResult<Vec<Certificate>> {
 }
 
 /// Reads a private key from a pkcs8 formatted, and possibly encoded file
-pub fn read_key_from_pkcs8(path: &Path) -> ProtoResult<PrivateKey> {
+pub fn read_key(path: &Path) -> ProtoResult<PrivateKey> {
     let mut file = BufReader::new(File::open(path)?);
 
-    let mut keys = match pkcs8_private_keys(&mut file) {
-        Ok(keys) => keys.into_iter().map(PrivateKey).collect::<Vec<_>>(),
-        Err(_) => {
-            return Err(ProtoError::from(format!(
-                "failed to read keys from: {}",
-                path.display()
-            )))
-        }
-    };
-
-    match keys.len() {
-        0 => return Err(format!("no keys available in: {}", path.display()).into()),
-        1 => (),
-        _ => warn!(
-            "ignoring other than the first key in file: {}",
-            path.display()
-        ),
+    loop {
+        match read_one(&mut file)? {
+            None => return Err(format!("no keys available in: {}", path.display()).into()),
```

Comment:
I think this would be easier to follow if you had the good cases first, and the error cases last.